File: THCS/2023_2024/NgheAn_Vinh_2324/Cau3/CHIAGIO.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <numeric>

using namespace std;

// Hàm kiểm tra có thể chia thành các nhóm có tổng là target_sum
bool canDivideGroups(const vector<int>& hours, int target_sum, vector<vector<int>>& groups) {
    int current_sum = 0;
    vector<int> temp_group;

    for (int hour : hours) {
        current_sum += hour;
        temp_group.push_back(hour);

        if (current_sum == target_sum) {
            groups.push_back(temp_group);
            current_sum = 0;
            temp_group.clear();
        } else if (current_sum > target_sum) {
            return false; // Không thể chia nhóm với target_sum này
        }
    }
    
    return current_sum == 0;
}

int main() {
    // Đọc dữ liệu từ file CHIAGIO.INP
    ifstream infile("CHIAGIO.INP");
    int n;
    infile >> n;

    vector<int> hours(n);
    for (int i = 0; i < n; i++) {
        infile >> hours[i];
    }
    infile.close();

    // Tính tổng tất cả các giờ làm việc
    int total = accumulate(hours.begin(), hours.end(), 0);

    // Tìm tổng S lớn nhất có thể chia thành nhiều nhóm nhất
    int max_k = 0;
    int best_sum = 0;
    vector<vector<int>> best_groups;

    for (int S = 1; S <= total; S++) {
        if (total % S == 0) {
            vector<vector<int>> groups;
            if (canDivideGroups(hours, S, groups) && groups.size() > max_k) {
                max_k = groups.size();
                best_groups = groups;
                best_sum = S;
            }
        }
    }

    // Ghi kết quả ra file CHIAGIO.OUT
    ofstream outfile("CHIAGIO.OUT");
    outfile << max_k << " " << best_sum << endl;
    for (const auto& group : best_groups) {
        for (int hour : group) {
            outfile << hour << " ";
        }
        outfile << endl;
    }
    outfile.close();

    return 0;
}

File: THCS/2023_2024/NgheAn_Vinh_2324/Cau3/CHIAGIO.py
# Đọc dữ liệu từ file CHIAGIO.INP
with open("CHIAGIO.INP", "r") as file:
    n = int(file.readline().strip())
    hours = list(map(int, file.readline().split()))

# Tính tổng các giờ làm việc
total = sum(hours)

# Hàm kiểm tra có thể chia thành nhóm với tổng là target_sum
def can_divide_groups(target_sum):
    current_sum = 0
    groups = []
    temp_group = []
    
    for hour in hours:
        current_sum += hour
        temp_group.append(hour)
        
        if current_sum == target_sum:
            groups.append(temp_group[:])
            current_sum = 0
            temp_group = []
        elif current_sum > target_sum:
            return False, []  # Không thể chia nhóm
        
    return current_sum == 0, groups

# Tìm tổng S lớn nhất có thể chia thành nhiều nhóm nhất
max_k = 0
best_groups = []
best_sum = 0

for S in range(1, total + 1):
    if total % S == 0:
        is_possible, groups = can_divide_groups(S)
        if is_possible and len(groups) > max_k:
            max_k = len(groups)
            best_groups = groups
            best_sum = S

# Ghi kết quả ra file CHIAGIO.OUT
with open("CHIAGIO.OUT", "w") as file:
    file.write(f"{max_k} {best_sum}\n")
    for group in best_groups:
        file.write(" ".join(map(str, group)) + "\n")
